### PR TITLE
IBP 3696 Move Environments to API Archictecture

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java [hunk 1]
```diff
@@ -25,35 +28,35 @@
...
 	}
 
 	@ApiOperation(value = "List all study instances with basic metadata.",
 		notes = "Returns list of all study instances with basic metadata.")
-	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES', 'INFORMATION_MANAGEMENT', 'BROWSE_STUDIES')")
+	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES')")
 	@RequestMapping(value = "/{cropname}/programs/{programUUID}/studies/{studyId}/instances", method = RequestMethod.GET)
 	@ResponseBody
-	public ResponseEntity<List<StudyInstance>> listStudyInstances(final @PathVariable String cropname, @PathVariable final String programUUID,
+	public ResponseEntity<List<StudyInstance>> listStudyEnvironments(final @PathVariable String cropname, @PathVariable final String programUUID,
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:59]: Use `listStudyInstances` to be consistent?

Comment (darla-leafnode) [reply src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:59]: Also why did we remove the following authorities from seeing the instances: 'INFORMATION_MANAGEMENT', 'BROWSE_STUDIES'

File: src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java [hunk 2]
```diff
@@ -64,9 +67,9 @@
 		return new ResponseEntity<>(studyInstances, HttpStatus.OK);
 	}
 
-	@ApiOperation(value = "Get study instance with basic metadata.",
-		notes = "Get study instance with basic metadata.")
-	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES', 'INFORMATION_MANAGEMENT', 'BROWSE_STUDIES')")
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:69]: Same here, why did we remove the following authorities from seeing the instance: 'INFORMATION_MANAGEMENT', 'BROWSE_STUDIES'

File: src/main/java/org/ibp/api/java/impl/middleware/common/MiddlewareFactory.java [hunk 3]
```diff
@@ -527,6 +528,12 @@ public ObservationUnitService getObservationUnitService() {
 		return new ObservationUnitServiceImpl(this.getCropDatabaseSessionProvider());
 	}
 
+	@Bean
+	@Scope(value = "request", proxyMode = ScopedProxyMode.TARGET_CLASS)
+	public ExperimentModelGenerator getExperimentModelGenerator() {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/MiddlewareFactory.java:533]: Why do we need `ExperimentModelGenerator ` in BMSAPI?

File: src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:59]: Shiela commented on the ticket that the API should only be accessible by a user with Manage Studies permission.

File: src/main/java/org/ibp/api/java/impl/middleware/common/MiddlewareFactory.java (hunk 3 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/common/MiddlewareFactory.java:533]: StudyInstanceServiceImpl needs it.

File: src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java (hunk 1 above)
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:59]: I think we should revert the old behavior to standardize the behavior of other GET endpoints in other study-related resource classes (eg. DatasetResource)

File: src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java [hunk 4]
```diff
@@ -75,4 +78,30 @@
 		return studyInstance.isPresent()? new ResponseEntity<>(studyInstance.get(), HttpStatus.OK) : new ResponseEntity(HttpStatus.NOT_FOUND);
 	}
 
+	@ApiOperation(value = "Add study instance data",
+		notes = "Add study environment data")
+	@PreAuthorize("hasAnyAuthority('ADMIN','BREEDING_ACTIVITIES','MANAGE_STUDIES')")
+	@RequestMapping(value = "/{cropname}/programs/{programUUID}/studies/{studyId}/instances/{instanceId}/instance-data", method = RequestMethod.POST)
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/study/StudyInstanceResource.java:84]: @darla-leafnode This is already merged but the name instance-data is not really friendly to someone that does not know the app. It does not describe a resource. Will need to find a more accurate name for this

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyInstanceServiceImpl.java [hunk 5]
```diff
@@ -25,70 +37,146 @@
...
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		instanceData.setInstanceDataId(instanceDataId);
+		instanceData.setInstanceId(instanceId);
+		this.middlewareStudyInstanceService.updateInstanceData(instanceData, isEnvironmentCondition);
+		return instanceData;
+	}
+
+	private Integer getEnvironmentDatasetId(final Integer studyId) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyInstanceServiceImpl.java:172]: @darla-leafnode This function should be partd of StudyService, it is the study responding by the datasets that it has.

**Pull request merged**